internal/controllers/phase_uninstall.go

```diff
@@ -0,0 +1,294 @@
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/openshift/reference-addon/internal/controllers/phase"
+	opsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"go.uber.org/multierr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func NewPhaseUninstall(signaler UninstallSignaler, uninstaller Uninstaller, opts ...PhaseUninstallOption) *PhaseUninstall {
+	var cfg PhaseUninstallConfig
+
+	cfg.Option(opts...)
+	cfg.Default()
+
+	return &PhaseUninstall{
+		cfg: cfg,
+
+		signaler:    signaler,
+		uninstaller: uninstaller,
+	}
+}
+
+type PhaseUninstall struct {
+	cfg PhaseUninstallConfig
+
+	signaler    UninstallSignaler
+	uninstaller Uninstaller
+}
+
+func (p *PhaseUninstall) Execute(ctx context.Context, req phase.Request) phase.Result {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	if !p.signaler.SignalUninstall(ctx) {
+		return phase.Success()
+	}
+
+	if err := p.uninstaller.Uninstall(ctx, p.cfg.AddonNamespace, p.cfg.OperatorName); err != nil {
+		return phase.Error(fmt.Errorf("uninstalling addon: %w", err))
+	}
+
+	return phase.Blocking()
+}
+
+type PhaseUninstallConfig struct {
+	Log logr.Logger
+
+	AddonNamespace string
+	OperatorName   string
+}
+
+func (c *PhaseUninstallConfig) Option(opts ...PhaseUninstallOption) {
+	for _, opt := range opts {
+		opt.ConfigurePhaseUninstall(c)
+	}
+}
+
+func (c *PhaseUninstallConfig) Default() {
+	if c.Log == nil {
+		c.Log = logr.Discard()
+	}
+}
+
+type PhaseUninstallOption interface {
+	ConfigurePhaseUninstall(*PhaseUninstallConfig)
+}
+
+type Uninstaller interface {
+	Uninstall(ctx context.Context, namespace, operatorName string) error
+}
+
+func NewUninstallerImpl(client client.Client, lister CSVLister, opts ...UninstallerImplOption) *UninstallerImpl {
+	var cfg UninstallerImplConfig
+
+	cfg.Option(opts...)
+	cfg.Default()
+
+	return &UninstallerImpl{
+		cfg: cfg,
+
+		client: client,
+		lister: lister,
+	}
+}
+
+type UninstallerImpl struct {
+	cfg UninstallerImplConfig
+
+	client client.Client
+	lister CSVLister
+}
+
+func (u UninstallerImpl) Uninstall(ctx context.Context, namespace, operatorName string) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	csvs, err := u.lister.ListCSVs(ctx, WithNamespace(namespace), WithPrefix(operatorName))
+	if err != nil {
+		return fmt.Errorf("listing ClusterServiceVersions with name %q: %w", operatorName, err)
+	}
+
+	var finalErr error
+
```

Comment:
does checking `len(csvs) == 0` make sense here. 
It could be interpretted as either:
- already uninstalled and uninstalltion is clean
- was expecting some but none found (error)

May be we can return finalErr nil only if `len(csvs) == 0`. When there are csvs to be deleted we can delete them and requeue the event to ensure that CSVs are gone in the next reconcile loop.